### PR TITLE
Fix a NullReferenceException in DocumentExtensions

### PR DESCRIPTION
**Pull request by aelij**
I encountered this while using the `Classifier`. This happens if you specify a span outside the the document. I wasn't sure if it should return the entire document's semantic model or throw an `ArgumentOutOfRangeException`; I chose the former.

Stack trace:

```
at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxFactsService.GetMemberBodySpanForSpeculativeBinding(SyntaxNode node)
at async Microsoft.CodeAnalysis.Shared.Extensions.DocumentExtensions.GetSemanticModelForNodeAsync(ISemanticModelService semanticModelService, ISyntaxFactsService syntaxFactService, Document document, SyntaxNode node, TextSpan span, CancellationToken cancellationToken)
at async Microsoft.CodeAnalysis.Shared.Extensions.DocumentExtensions.GetSemanticModelForSpanAsync(?)
at async Microsoft.CodeAnalysis.Classification.Classifier.GetClassifiedSpansAsync(?)
```

### CODE REVIEW

**Comment by Pilchie** [comment]
Hi @aelij,

Thanks for the submission!  However, I think I personally would prefer to have this throw.  It's generally a programming error to ask for a span outside the span of the document.  Did you run into this with a regular file in Visual Studio, or did was it when you were using the API?

Anyone else on @dotnet/roslyn-ide want to comment?  @mattwar?

File: src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs [hunk 1]
```diff
@@ -59,6 +59,10 @@ public static bool IsOpen(this Document document)
                 }
 
                 var node = token.Parent.AncestorsAndSelf().FirstOrDefault(a => a.FullSpan.Contains(span));
+                if (node == null)
+                {
+                    return await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                }
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: Style Nit: We generally prefer a blank line after a `}`.

File: src/Workspaces/CoreTest/ExtensionsTests/DocumentExtensionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+namespace Microsoft.CodeAnalysis.UnitTests.ExtensionsTests
+{
+    public class DocumentExtensionsTests
+    {
+        private readonly AdhocWorkspace _ws = new AdhocWorkspace();
+        private readonly Project _emptyProject;
+
+        public DocumentExtensionsTests()
+        {
+            _emptyProject = _ws.AddProject(
```
Comment (Pilchie) [inline src/Workspaces/CoreTest/ExtensionsTests/DocumentExtensionsTests.cs]: Take a look at the `TestWorkspace` class.  It makes it simple to construct a workspace with the right services, and specify spans/positions within the documents.

**Comment by jasonmalinowski** [comment]
Agreed -- this should throw in that case. (I wonder what bugs we'll uncover with that!)

**Comment by aelij** [comment]
Is `throw new ArgumentOutOfRange(nameof(span))` enough or should I add some message?

@Pilchie I ran into this while using the API to hook up syntax highlighting in AvalonEdit.

File: src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs [hunk 3]
```diff
@@ -59,6 +59,10 @@ public static bool IsOpen(this Document document)
                 }
 
                 var node = token.Parent.AncestorsAndSelf().FirstOrDefault(a => a.FullSpan.Contains(span));
+                if (node == null)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: Hrmmm... how could node be null?  When would we have a span that wasn't contained in the full span of some node in the document?

**Comment by CyrusNajmabadi** [comment]
I would "throw an ArgumentOutOfRangeException".  The span is not within bounds.  There is something very wrong with the caller.

File: src/Workspaces/CoreTest/ExtensionsTests/DocumentExtensionsTests.cs (hunk 2 above)
Comment (aelij) [inline src/Workspaces/CoreTest/ExtensionsTests/DocumentExtensionsTests.cs]: Thanks, @Pilchie, but the `TestWorkspace` in the `ServicesTest` project doesn't seem to do that. There's one in `EditorServicesTest`, but I'm assuming referencing it is not an option. I copied this from another test in the same project.

Comment (Pilchie) [inline src/Workspaces/CoreTest/ExtensionsTests/DocumentExtensionsTests.cs]: Doh, you're right of course.  We really need to move the xml consumption stuff down to workspaces and unify the test workspaces.

**Comment by aelij** [comment]
In case it got missed - I pushed the requested changes to my branch.

**Comment by davkean** [comment]
@dotnet/roslyn-ide Can you take another look at it?

File: src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs [hunk 4]
```diff
@@ -59,6 +59,11 @@ public static bool IsOpen(this Document document)
...
+                {
+                    throw new ArgumentOutOfRangeException(nameof(span));
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: Seems more reasonable, should we just do if (!root.FullSpan.Contains(span)) earlier and throw instead? This seems to be a fairly circuitous way to check the simpler case.

**Comment by DustinCampbell** [comment]
@Pilchie, @CyrusNajmabadi, @jasonmalinowski: What's the status on this PR?

**Comment by jasonmalinowski** [comment]
I guess we are waiting for @aelij to respond to my suggestion?

**Comment by aelij** [comment]
Oh, I was waiting for something actionable... :)

I'd leave it as is, since obviously this isn't a common code path, so it's a bit of a waste to check the span's range on every call.

**Comment by jaredpar** [comment]
We apologize, but we are closing this PR due to code drift. We regret letting this PR go so long without attention, but at this point the code base has changed too much for us to revisit older PRs. Going forward, we will manage PRs better under an SLA currently under draft in issue #26266 – we encourage you to add comments to that draft as you see fit.   Although that SLA is still in draft form, we nevertheless want to move forward with the identification of older PRs to give contributors as much early notice as possible to review and update them.

If you are interested in pursuing this PR, please reset it against the head of master and we will reopen it. Thank you!

**Pull request closed without merging**